core/src/main/java/com/linecorp/armeria/server/management/HeapDumpService.java

```diff
@@ -0,0 +1,185 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License
+ */
+/*
+ * Copyright 2012-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.linecorp.armeria.server.management;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.lang.management.ManagementFactory;
+import java.lang.management.PlatformManagedObject;
+import java.nio.file.Files;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import javax.annotation.Nullable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.common.ContentDisposition;
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.QueryParams;
+import com.linecorp.armeria.server.HttpService;
+import com.linecorp.armeria.server.RequestTimeoutException;
+import com.linecorp.armeria.server.ServiceRequestContext;
+import com.linecorp.armeria.server.file.HttpFile;
+
+enum HeapDumpService implements HttpService {
+
+    // Forked from https://github.com/spring-projects/spring-boot/blob/531ee83c6ae42d8027a5241faf42c2015b1a031c/spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/management/HeapDumpWebEndpoint.java
+
+    INSTANCE;
+
+    private static final Logger logger = LoggerFactory.getLogger(HeapDumpService.class);
+
+    private final Lock lock = new ReentrantLock();
+
+    @Nullable
+    private HeapDumper heapDumper;
+
+    @Nullable
+    private Throwable unavailabilityCause;
+
+    @Override
+    public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
+        if (unavailabilityCause != null) {
+            return HttpResponse.ofFailure(unavailabilityCause);
+        }
+
+        final CompletableFuture<HttpResponse> responseFuture = new CompletableFuture<>();
+        ctx.blockingTaskExecutor().execute(() -> {
```

Comment:
I'm somewhat concerned about this task being blocked by other blocking tasks, because this sort of operation is usually performed when something's wrong. How about creating a new thread instead?